### PR TITLE
Some matrix optimizations

### PR DESCRIPTION
**Pull request by joschmitt**
Including and extending the work of @fieker.

I add tests once I see the CI result as usual. ~(Which needs an AbstractAlgebra update; I forgot.)~

### CODE REVIEW

File: src/flint/fq_default_mat.jl [hunk 1]
```diff
@@ -78,6 +78,25 @@ end
...
 
+function setindex!(a::FqMatrix, b::FqMatrix, r::UnitRange{Int64}, c::UnitRange{Int64})
+  _checkbounds(a, r, c)
+  size(b) == (length(r), length(c)) || throw(DimensionMismatch("tried to assign a $(size(b, 1))x$(size(b, 2)) matrix to a $(length(r))x$(length(c)) destination"))
+  A = view(a, r, c)
+  ccall((:fq_default_mat_set, libflint), Nothing,
+        (Ref{FqMatrix}, Ref{FqMatrix}, Ref{FqField}), A, b, base_ring(a))
+end
+
+function Generic.add_one!(a::FqMatrix, i::Int, j::Int)
```
[the remark this answers is not in the document]
Comment (thofma) [reply src/flint/fq_default_mat.jl:89]: There is `fq_default_sub_one`, but no `fq_default_add_one`. So you could do a `neg!(sub_one!(neg!(x)))` to do it allocation free. Not sure if this is a good idea.

Any ideas @fieker?

**Comment by fieker** [comment]
On Thu, Apr 11, 2024 at 07:12:36AM -0700, Johannes Schmitt wrote:
> @joschmitt commented on this pull request.
> 
> 
> 
> > @@ -78,6 +78,25 @@ end
>  setindex!(a::FqMatrix, u::Integer, i::Int, j::Int) =
>          setindex!(a, base_ring(a)(u), i, j)
>  
> +function setindex!(a::FqMatrix, b::FqMatrix, r::UnitRange{Int64}, c::UnitRange{Int64})
> +  _checkbounds(a, r, c)
> +  size(b) == (length(r), length(c)) || throw(DimensionMismatch("tried to assign a $(size(b, 1))x$(size(b, 2)) matrix to a $(length(r))x$(length(c)) destination"))
> +  A = view(a, r, c)
> +  ccall((:fq_default_mat_set, libflint), Nothing,
> +        (Ref{FqMatrix}, Ref{FqMatrix}, Ref{FqField}), A, b, base_ring(a))
> +end
> +
> +function Generic.add_one!(a::FqMatrix, i::Int, j::Int)
> 
> Should this function be added for all the different flint matrix types? And is there a better way to do it?
> With the pointer it needs 1 allocation (I assume for `one(F)`) in comparison to 3 allocations for the naive `a[i, j] += 1`. 
> 
add_one is intended to not do any alloc. If it does, then my code is
stupid. Almost all flint rings allow add_si or similar to not create the
ring elem first.
Asymptotically, it should be for all types, however, for nmod disguised
as Fq the generic fallback  is particularly bad.
Furthermore, the only code calling it is from the inj/proj mat stuff...
I think

**Comment by thofma** [comment]
`fq_default` does not have the `add_si` version.

**Comment by fieker** [comment]
On Thu, Apr 11, 2024 at 07:40:35AM -0700, Tommy Hofmann wrote:
> `fq_default` does not have the `add_si` version.
but the nmod does I think. IN particular since 1 is a valid nmod.

**Comment by joschmitt** [comment]
I added `add_one!` for the flint matrix types. It is allocation free except for `fq_mat` and `fq_nmod_mat` (and the corresponding cases in `fq_default_mat`). For these, `neg(sub_one(neg(...)))` is better than `add(..., one)`, but still needs one allocation.

File: src/flint/fmpz_mod_mat.jl [hunk 2]
```diff
@@ -280,6 +288,20 @@ function mul!(z::Vector{ZZRingElem}, a::Vector{ZZRingElem}, b::T) where T <: Zmo
...
 end
 
+function Generic.add_one!(a::ZZModMatrix, i::Int, j::Int)
+  @boundscheck Generic._checkbounds(a, i, j)
+  GC.@preserve a begin
+    x = mat_entry_ptr(a, i, j)
+    ccall((:fmpz_mod_add_si, libflint), Nothing,
+          (Ptr{ZZModRingElem}, Ptr{ZZModRingElem}, Int, Ref{ZZModRing}),
+          x, x, 1, base_ring(a))
+    ccall((:fmpz_mod, libflint), Nothing,
```
Comment (thofma) [inline src/flint/fmpz_mod_mat.jl:298]: Are you sure this is necessary? Looking at https://github.com/flintlib/flint/blob/flint-2.9/fmpz_mod/add_fmpz.c#L42, it seems that `fmpz_mod_add_si` does a reduction.

Comment (joschmitt) [reply src/flint/fmpz_mod_mat.jl:298]: I was actually wondering about this:
```
julia> R, _ = residue_ring(ZZ, ZZ(3));

julia> a = zero_matrix(R, 2, 2);

julia> a[1, 1] = 2;

julia> i = 1; j = 1;

julia> GC.@preserve a begin
           x = Nemo.mat_entry_ptr(a, i, j)
           ccall((:fmpz_mod_add_si, Nemo.libflint), Nothing,
                 (Ptr{ZZModRingElem}, Ptr{ZZModRingElem}, Int, Ref{ZZModRing}),
                 x, x, 1, base_ring(a))
       end

julia> a
[3   0]
[0   0]

```

Comment (joschmitt) [reply src/flint/fmpz_mod_mat.jl:298]: I don't understand this function.
```
julia> R, _ = residue_ring(ZZ, ZZ(3));

julia> x = R(); y = R();

julia> ccall((:fmpz_mod_add_si, Nemo.libflint), Nothing,
                        (Ref{ZZModRingElem}, Ref{ZZModRingElem}, Int, Ref{ZZModRing}),
                        y, x, 1, R)

julia> y
0
```

**Comment by joschmitt** [comment]
I corrected the `fmpz_mod` cases as discussed, so this is good to go now.

**Pull request merged**